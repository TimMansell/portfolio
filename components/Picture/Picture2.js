export const Picture = () => {
  return <>test</>;
};

export default Picture;
