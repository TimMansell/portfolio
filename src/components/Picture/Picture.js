import React from 'react';
import PropTypes from 'prop-types';

import styles from './Picture.module.scss';

export const Picture = ({ srcs, name, defaultImg }) => {
  const fallbackImg = defaultImg || srcs[0].src;

  return (
    <picture title={name} data-e2e="picture">
      {srcs.map(({ src, format }, index) => (
        <source
          type={`image/${format}`}
          srcSet={src}
          key={index}
          data-test={`picture-source-${index}`}
        />
      ))}

      <img
        className={styles.img}
        alt={name}
        src={fallbackImg}
        data-test="picture-img"
      />
    </picture>
  );
};

Picture.propTypes = {
  srcs: PropTypes.arrayOf(
    PropTypes.shape({
      src: PropTypes.string.isRequired,
      format: PropTypes.string.isRequired,
    })
  ).isRequired,
  name: PropTypes.string.isRequired,
  defaultImg: PropTypes.string,
};

export default Picture;
