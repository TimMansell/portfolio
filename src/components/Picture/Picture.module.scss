.img {
  max-width: 100%;
  height: auto;
  object-fit: cover;
}

.fullscreen {
  height: 100%;
}
