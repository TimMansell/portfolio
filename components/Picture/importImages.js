import { getSources, getFallbackImage } from './helpers';

export const importImages = (src, srcSizes, types) => {
  const sources = types.flatMap(getSources(src, srcSizes));
  const fallbackImg = getFallbackImage(src, srcSizes, types);

  return {
    sources,
    fallbackImg,
  };
};
