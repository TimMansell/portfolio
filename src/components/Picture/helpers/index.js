import { getSizes, getFilePath } from './importImages';

export const getSources = (src, srcSizes) => (type) =>
  srcSizes.map(getSizes(src, type));

export const getFallbackImage = (src, srcSizes, types) => {
  const [fallbackSrc] = [...srcSizes].reverse();
  const { sizes } = fallbackSrc;
  const [size] = [...sizes].reverse();
  const [type] = [...types].reverse();

  const filePath = getFilePath(src, size, type);

  const fallbackImg = new URL(`../../${filePath}`, import.meta.url).href;

  return fallbackImg;
};
