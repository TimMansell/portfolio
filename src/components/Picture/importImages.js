import { getSources, getFallbackImage } from './helpers';

export const importImages = (src, types, srcSizes) => {
  const sources = types.map(getSources(src, srcSizes));
  const fallbackImg = getFallbackImage(src, srcSizes, types);

  // console.log(JSON.stringify(sources));

  return {
    sources: sources[0],
    fallbackImg,
  };
};
