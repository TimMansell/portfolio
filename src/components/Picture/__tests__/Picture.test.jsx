import { it, describe, expect } from 'vitest';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Picture from '../Picture';

const url = import.meta.url.split('/').slice(0, -3).join('/');

const props = {
  title: 'test title',
  types: ['avif', 'webp', 'jpg'],
  src: 'img/test',
  srcSizes: [{ sizes: ['480', '640'] }],
  width: '50',
};

describe('Picture', () => {
  it('should render my component', () => {
    // eslint-disable-next-line
    const wrapper = shallow(<Picture {...props} />);
  });

  it.skip('should match snapshot', () => {
    const snapshot = renderer.create(<Picture {...props} />).toJSON();

    expect(snapshot).toMatchSnapshot();
  });

  it('should render fullscreen image', () => {
    const wrapper = shallow(<Picture {...props} isFullscreen />);
    const image = wrapper.find('[data-test="picture-img"]');

    expect(image.hasClass('fullscreen')).toBeTruthy();
  });

  it('should render non-fullscreen image', () => {
    const wrapper = shallow(<Picture {...props} />);
    const image = wrapper.find('[data-test="picture-img"]');

    expect(image.hasClass('fullscreen')).toBeFalsy();
  });

  it('should render lazy loaded image', () => {
    const wrapper = shallow(<Picture {...props} isLazy />);
    const image = wrapper.find('[data-test="picture-img"]');

    expect(image.props().loading).toBe('lazy');
  });

  it('should render non lazy loaded image', () => {
    const wrapper = shallow(<Picture {...props} />);
    const image = wrapper.find('[data-test="picture-img"]');

    expect(image.props().loading).toBe('auto');
  });

  it('should have correct width', () => {
    const wrapper = shallow(<Picture {...props} />);
    const image = wrapper.find('[data-test="picture-img"]');

    expect(image.props().width).toBe(props.width);
  });

  it('should have correct <source> properties', () => {
    const wrapper = shallow(<Picture {...props} />);
    const source = wrapper.find('[data-test="picture-source-0"]');

    expect(source.props().srcSet).toBe(
      `${url}/img/avif/test-480.avif 480w, ${url}/img/avif/test-640.avif 640w`
    );
    expect(source.props().type).toBe(`image/${props.types[0]}`);
  });

  it('should have correct number of <source>', () => {
    const wrapper = shallow(<Picture {...props} />);
    const sources = wrapper.find('[data-test*="picture-source"]');

    expect(sources).toHaveLength(3);
  });

  it('should render fallback in <img>', () => {
    const wrapper = shallow(<Picture {...props} />);
    const source = wrapper.find('[data-test="picture-img"]');

    expect(source.props().src).toBe(`${url}/img/jpg/test-640.jpg`);
    expect(source.props().alt).toBe(props.title);
  });
});
