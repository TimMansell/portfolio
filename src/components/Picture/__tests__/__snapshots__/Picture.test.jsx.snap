// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Picture > should match snapshot 1`] = `
<picture
  data-e2e="picture"
  title="test title"
>
  <source
    data-test="picture-source-0"
    srcSet="file:///H:/projects/portfolio/src/components/Picture/__tests__/img/avif/test-480.avif 480w, file:///H:/projects/portfolio/src/components/Picture/__tests__/img/avif/test-640.avif 640w"
    type="image/avif"
  />
  <source
    data-test="picture-source-1"
    srcSet="file:///H:/projects/portfolio/src/components/Picture/__tests__/img/webp/test-480.webp 480w, file:///H:/projects/portfolio/src/components/Picture/__tests__/img/webp/test-640.webp 640w"
    type="image/webp"
  />
  <source
    data-test="picture-source-2"
    srcSet="file:///H:/projects/portfolio/src/components/Picture/__tests__/img/jpg/test-480.jpg 480w, file:///H:/projects/portfolio/src/components/Picture/__tests__/img/jpg/test-640.jpg 640w"
    type="image/jpg"
  />
  <img
    alt="test title"
    className="img"
    data-e2e="picture-img"
    data-test="picture-img"
    loading="auto"
    src="file:///H:/projects/portfolio/src/components/Picture/__tests__/img/jpg/test-640.jpg"
    width="50"
  />
</picture>
`;
